attestation-agent/attester/src/cca/mod.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2023 Arm Ltd.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use super::Attester;
+use anyhow::*;
+use base64::Engine;
+use nix::fcntl::{open, OFlag};
+use nix::sys::stat::Mode;
+use nix::unistd::close;
+use serde::{Deserialize, Serialize};
+use std::path::Path;
+
+const CCA_DEVICE_PATH: &str = "/dev/cca_attestation";
+
+// NOTE: The path might be different when the CCA feature is public available, will come back to update the actual path if needed.
+pub fn detect_platform() -> bool {
+    Path::new(CCA_DEVICE_PATH).exists()
+}
+
+#[derive(Debug, Default)]
+pub struct CCAAttester {}
+
+#[derive(Serialize, Deserialize)]
+struct CcaEvidence {
+    /// CCA token
+    token: Vec<u8>,
+}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct cca_ioctl_request {
+    challenge: [u8; 64],
+    token: [u8; 4096],
+    token_length: u64,
+}
+
+nix::ioctl_readwrite!(cca_attestation_request, b'A', 1, cca_ioctl_request);
+
+#[async_trait::async_trait]
+impl Attester for CCAAttester {
+    async fn get_evidence(&self, mut challenge: Vec<u8>) -> Result<String> {
+        challenge.resize(64, 0);
+        let token = attestation(&challenge)?;
+        let evidence = CcaEvidence { token };
+        let ev = serde_json::to_string(&evidence).context("Serialize CCA evidence failed")?;
+        Ok(ev)
+    }
+}
+
+fn attestation(challenge: &[u8]) -> Result<Vec<u8>, Error> {
+    log::info!("cca_test::attestation started");
+
+    match open(CCA_DEVICE_PATH, OFlag::empty(), Mode::empty()) {
```

Comment:
It might be better to use a`?` and a `context` here rather than have such a big match case.

Reply:
thanks for the suggestion, I don't have strong opinion about this. If this is something must be updated, pls just let me know, thanks!

Reply:
one advantage for the current implementation is that it's easier for me to close the file descriptor on error.